extensions/markdown-language-features/src/test/markdownLink.test.ts

```diff
@@ -5,128 +5,215 @@
+			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
 		});
 
-		test('Should evaluate updateTitle as true for pasting over a Markdown image link', () => {
-			const smartPaste = checkSmartPaste("![a](bc)", 0, 8);
+		const linkSkinnyDoc: SkinnyTextDocument = {
+			uri: vscode.Uri.file('/path/to/your/file'),
+			offsetAt: function () { return 0; },
+			getText: function () { return '[a](bcdef)'; },
+		};
+
+		test('Should evaluate updateTitle as true for pasting over a Markdown link', () => {
+			const range = new vscode.Range(0, 0, 0, 10);
+			const smartPaste = checkSmartPaste(linkSkinnyDoc, range);
 			assert.strictEqual(smartPaste.updateTitle, true);
 		});
 
 		test('Should evaluate pasteAsMarkdownLink as false for pasting within a Markdown link', () => {
-			const smartPaste = checkSmartPaste("[a](bcdef)", 4, 6);
+			const range = new vscode.Range(0, 4, 0, 6);
+			const smartPaste = checkSmartPaste(linkSkinnyDoc, range);
+
 			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
 		});
 
-		test('Should evaluate pasteAsMarkdownLink as false for pasting within a Markdown image link', () => {
-			const smartPaste = checkSmartPaste('![alt](https://)', 7, 15);
-			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
+
+		const imageLinkSkinnyDoc: SkinnyTextDocument = {
+			uri: vscode.Uri.file('/path/to/your/file'),
+			offsetAt: function () { return 0; },
+			getText: function () { return '![a](bcdef)'; },
+		};
+
+		test('Should evaluate updateTitle as true for pasting over a Markdown image link', () => {
+			const range = new vscode.Range(0, 0, 0, 11);
+			const smartPaste = checkSmartPaste(imageLinkSkinnyDoc, range);
+			assert.strictEqual(smartPaste.updateTitle, true);
 		});
 
-		test('Should evaluate pasteAsMarkdownLink as false for pasting within a code block', () => {
-			const smartPaste = checkSmartPaste('```\r\n\r\n```', 5, 5);
+		test('Should evaluate pasteAsMarkdownLink as false for pasting within a Markdown image link', () => {
+			const range = new vscode.Range(0, 5, 0, 10);
+			const smartPaste = checkSmartPaste(imageLinkSkinnyDoc, range);
 			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
 		});
 
+		const inlineCodeSkinnyCode: SkinnyTextDocument = {
+			uri: vscode.Uri.file('/path/to/your/file'),
+			offsetAt: function () { return 0; },
+			getText: function () { return '``'; },
+		};
+
 		test('Should evaluate pasteAsMarkdownLink as false for pasting within inline code', () => {
-			const smartPaste = checkSmartPaste('``', 1, 1);
+			const range = new vscode.Range(0, 1, 0, 1);
+			const smartPaste = checkSmartPaste(inlineCodeSkinnyCode, range);
 			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
 		});
 
-		test('Should evaluate pasteAsMarkdownLink as false for pasting within a math block', () => {
-			const smartPaste = checkSmartPaste('$$$\r\n\r\n$$$', 5, 5);
-			assert.strictEqual(smartPaste.pasteAsMarkdownLink, false);
-		});
+		const inlineMathSkinnyDoc: SkinnyTextDocument = {
```

Comment:
Try moving constants used inside the test into the test itself. This makes them easier to read

Reply:
You may also want to create a helper function that creates `SkinnyTextDocuments` 

Or even better, add a new class that implements `SkinnyTextDocuments` for your tests. That way you can just call `new InMemoryDocument('doc text')` or something similar